js/blob.ts

```diff
@@ -2,13 +2,64 @@
 import * as domTypes from "./dom_types";
 import { containsOnlyASCII, hasOwnProperty } from "./util";
 import { TextEncoder } from "./text_encoding";
+import { build } from "./build";
 
 export const bytesSymbol = Symbol("bytes");
 
 function convertLineEndingsToNative(s: string): string {
-  // TODO(qti3e) Implement convertLineEndingsToNative.
-  // https://w3c.github.io/FileAPI/#convert-line-endings-to-native
-  return s;
+  let nativeLineEnd = "\n";
+  if (build.os == "win") {
+    nativeLineEnd = "\r\n";
+  }
+
+  let result = "";
+  let position = 0;
+
+  let collectionResult = collectSequenceNotCRLF(s, position);
+
+  let token = collectionResult.collected;
+  position = collectionResult.newPosition;
+
+  result = result + token;
+
+  while (position < s.length) {
+    let c = s.charAt(position);
+    if (c == "\r") {
+      result = result + nativeLineEnd;
+      position++;
+      if (position < s.length && s.charAt(position) == "\n") {
+        position++;
+      }
+    } else if (c == "\n") {
+      position++;
+      result = result + nativeLineEnd;
+    }
+
+    collectionResult = collectSequenceNotCRLF(s, position);
+
+    token = collectionResult.collected;
+    position = collectionResult.newPosition;
+
+    result = result + token;
+  }
+
+  return result;
+}
+
+function collectSequenceNotCRLF(
+  s: string,
+  position: number
+): { collected: string; newPosition: number } {
+  let result = "";
+  while (position < s.length) {
+    let c = s.charAt(position);
+    if (c == "\r" || c == "\n") {
+      break;
+    }
+    result = result.concat(c);
```

Comment:
There is no need to do this call on every iteration, you can obtain the same result using `slice()` at the end.
```ts
const start = position;
while (...) { ... }
return { collected: s.slice(start, position), newPosition: position };
```

I would probably use something like this:
```ts
function collectSequenceNotCRLF(s, position) {
  const start = position;

  for (
    let c = s.charAt(position);
    position < s.length && !(c == '\r' || c == '\n');
    c = s.charAt(++position)
  );

  return { newPosition: position, collected: s.slice(start, position) };
}
```